source/receiver.cpp

```diff
@@ -0,0 +1,127 @@
+#include "receiver.hpp"
+
+#include <spdlog/spdlog.h>
+#include <iostream>
+
+#include "event.hpp"
+#include "flow.hpp"
+#include "link.hpp"
+#include "scheduler.hpp"
+
+namespace sim {
+
+Receiver::Receiver() : m_router(std::make_unique<RoutingModule>()) {}
+
+bool Receiver::add_inlink(std::shared_ptr<ILink> link) {
+    if (link == nullptr) {
+        spdlog::warn("Passed link is null");
+        return false;
+    }
+
+    if (this != link->get_to().get()) {
+        spdlog::warn("Link destination device is incorrect (expected current device)");
+        return false;
+    }
+    m_router->add_inlink(link);
+    return true;
+}
+
+bool Receiver::update_routing_table(std::shared_ptr<IRoutingDevice> dest,
+                                     std::shared_ptr<ILink> link) {
+    if (link == nullptr) {
+        spdlog::warn("Passed link is null");
+        return false;
+    }
+
+    if (dest == nullptr) {
+        spdlog::warn("Passed destination is null");
+        return false;
+    }
+
+    if (this != link->get_from().get()) {
+        spdlog::warn("Link source device is incorrect (expected current device)");
+        return false;
+    }
+    m_router->update_routing_table(dest, link);
+    return true;
+}
+
+std::vector<std::shared_ptr<IRoutingDevice>> Receiver::get_neighbours() const {
+    return m_router->get_neighbours();
+};
+
+std::shared_ptr<ILink> Receiver::next_inlink() {
+    return m_router->next_inlink();
+};
+
+std::shared_ptr<ILink> Receiver::get_link_to_destination(
+    std::shared_ptr<IRoutingDevice> dest) const {
+        return m_router->get_link_to_destination(dest);
+};
+
+DeviceType Receiver::get_type() const {
+    return DeviceType::RECEIVER;
+}
+
+std::uint32_t Receiver::process() {
+    std::shared_ptr<ILink> current_inlink = m_router->next_inlink();
+    std::uint32_t total_processing_time = 1;
+
+    if (current_inlink == nullptr) {
+        spdlog::warn("No available inlinks for device");
+        return total_processing_time;
+    }
+
+    std::optional<Packet> opt_data_packet = current_inlink->get_packet();
+    if (!opt_data_packet.has_value()) {
+        spdlog::warn("No available packets from inlink for device");
+        return total_processing_time;
+    }
+
+    Packet data_packet = opt_data_packet.value();
+    if (data_packet.flow == nullptr) {
+        // TODO: discuss do we need to discard packet in such scenario or process it
+        spdlog::error("Packet flow does not exist");
+        return total_processing_time;
+    }
+
+    std::shared_ptr<IReceiver> receiver = data_packet.flow->get_receiver();
+    if (data_packet.type == DATA && receiver.get() == this) {

```

Comment:
Please, use `get_destination` you implemented:
```suggestion
    if (data_packet.get_destination().get() == this) {

```